### ISSUE TITLE
Linting and formatting CSS

### ISSUE THREAD

**Issue by gavinorland** [survivejs/maintenance-book#49]
Maybe a bit more could be said about the current state of play between Stylelint, Stylefmt and Prettier when it comes to formatting CSS. I mean the case use for these tools - what makes one worth using alongside, or instead of, another. I have been trying to find out, for example:

- How many rules does [Prettier](https://github.com/prettier/prettier) implement over CSS? Some are exposed via options, I guess there are others.
- Will Prettier support plugins, as [Stylelint](https://github.com/stylelint/stylelint) does (such as [Order](https://github.com/hudochenkov/stylelint-order))? This strikes me as a very useful plugin which makes Stylelint worth using right away.
- Can Stylelint `--fix` all the rules that [Stylefmt](https://github.com/morishitter/stylefmt) used to be able to do (thus rendering the latter definitely redundant)?
- Are there things that Prettier fixes in terms of formatting that Stylelint does not, thus making it worth using?

Just as we can [run Prettier for JS via ESLint](https://github.com/prettier/eslint-plugin-prettier), there is also [a tool](https://github.com/hugomrdias/prettier-stylelint) for running Prettier with Stylelint. It's a bit different though: apparently Prettier is run directly (Stylelint doesn't apply the formatting) then Stylelint is run. The tool just tries to match Prettier options with Stylelint rules.

Bottom line: what's the best workflow with these tools to take maximum advantage of linting and fixing, without redundancy?

**Comment by sapegin** [comment]
> How many rules does Prettier implement over CSS? Some are exposed via options, I guess there are others.

It doesn‘t implement any rules, it formats code, so you’ll need to remove any code style rules from Stylelint config. Same as with ESLint.

> Will Prettier support plugins, as Stylelint does (such as Order)? This strikes me as a very useful plugin which makes Stylelint worth using right away.

I’m 99.9% sure it won’t. The whole point of Prettier is no config and no arguments on code style in teams ;-) But you can continue use Stylelint with this plugin.

> Can Stylelint --fix all the rules that Stylefmt used to be able to do (thus rendering the latter definitely redundant)?

That’s why we removed Stylefmt from the book ;-)

> Are there things that Prettier fixes in terms of formatting that Stylelint does not, thus making it worth using?

Stylelint doesn’t reformat code, it just enforces some code style rules. I think I’m answering this question [here](https://medium.freecodecamp.org/why-robots-should-format-our-code-159fd06d17f7).

Should we all these to the book? ;-)

**Comment by sapegin** [comment]
> Bottom line: what's the best workflow with these tools to take maximum advantage of linting and fixing, without redundancy?

The one we explain in the book ;-)

**Comment by gavinorland** [comment]
What I meant by Prettier implementing "rules" was its internal rules for formatting which might  match ESLint's. Sorry if I didn't make that clear.

I agree it looks very unlikely that Prettier will implement plugins - I appreciate it is deliberately minimal in config.

> That’s why we removed Stylefmt from the book ;-)

I'll take that as a "Yes" that Stylelint can now fix all the rules that Stylefmt could - thanks!

> Stylelint doesn’t reformat code

Couldn't we say Stylelint does some reformatting in the sense of using `--fix` to change [spacing](https://github.com/prettier/prettier#how-does-it-compare-to-eslint-or-tslint-stylelint) etc.?

Thanks - I will have another look at your article, although I think it relates only to JS. I was just wondering if there is formatting of CSS that Prettier performs that `stylelint --fix` cannot.

Yes, as you say in the book, perhaps ESLint and Prettier are both worth running. Feel free to close this if you think you already have these points sufficiently covered.

**Comment by sapegin** [comment]
> What I meant by Prettier implementing "rules" was its internal rules for formatting which might match ESLint's. Sorry if I didn't make that clear.

I don’t think such comparison makes much sense because Prettier completely reformats the code from AST and only very basic rules like quotes or semicolons may match ESLint.

> Couldn't we say Stylelint does some reformatting in the sense of using --fix to change spacing etc.?

We have some terminology issues here ;-) ESLint / Stylelint do change formatting a bit: add or remove whitespace, change type of quotes, etc. But they don’t *reprint* code from AST like Prettier does. So ESLint / Stylelint try to *fix* already formatted code and Prettier completely (not 100% but very close) ignores original code formatting. I don’t think we should use the same word to describe both processes. @bebraw what do you think? ;-)

> Thanks - I will have another look at your article, although I think it relates only to JS. I was just wondering if there is formatting of CSS that Prettier performs that stylelint --fix cannot.

It describes the difference between two different processes: Stylelint can‘t do the thing we don’t know how to call that Prettier does ;-)

> Yes, as you say in the book, perhaps ESLint and Prettier are both worth running. Feel free to close this if you think you already have these points sufficiently covered.

I think it can be improved, so let’s keep it open 🍕

**Comment by gavinorland** [comment]
I would certainly say that ESLint and Stylelint format code when they fix those kinds of rules. By contrast, Prettier _reprints_ all of the code, as you say, also formatting. It's just more of a "nuclear option"!

Just to be perfectly clear, is there anything that Prettier can format that `stylelint --fix` cannot?

I'll get back to this tomorrow and mention if there is anything else I'm still not clear on. Thanks again :)

**Comment by gavinorland** [comment]
A [new article on this topic](https://css-tricks.com/prettier-stylelint-writing-clean-css-keeping-clean-code-two-tool-game/) with an interesting post by the main developer of stylelint. Seems to me the main thing Prettier fixes (reprints) that Stylelint doesn't is line length. Main benefit of using stylelint in tandem with Prettier seems to be ability to use plugins like Order, as I mentioned (though it lints many other things too). Stylelint cannot yet fix all the rules that stylefmt can (see for example `declaration-colon-space-after`), but they're adding them all the time. :)

**Comment by sapegin** [comment]
Probably we need to add a short section on liters v. formatters to clarify why you need to use both.

**Comment by sapegin** [comment]
This should make it more clear: #62

**Comment by bebraw** [comment]
@gavinorland What do you think? Time to close?

**Comment by gavinorland** [comment]
Looks good but I think I might be able to add something. Please leave it open and I'll get back to you ASAP.

**Comment by gavinorland** [comment]
Right, I was just reviewing what exactly are the differences between the tools again. Of course, Prettier actually reformats, but I wanted to check on what it does that ESLint cannot. If you wanted to be a bit more explicit for the curious reader (like me!) maybe this would be of use:

> A common misconception is that if you use a code formatter like Prettier then you don’t need a linter anymore. However, Prettier will not touch any code quality issues, such as unused variables or implicit globals. And while both ESLint and Prettier can auto-fix semi-colon, indent, quote, comma dangling and spacing preferences, Prettier can operate on more file types (including JSON and CSS) and, as it reprints, it can also for example auto-fix line lengths where possible and remove unnecessary space between code blocks. Therefore at present it's worth using both of these tools together.

Up to you guys!

**Comment by sapegin** [comment]
@gavinorland Take a look at #63, is it better now?
